File: flutter/lib/core/logic/markdown_document.dart
import 'package:markdown/markdown.dart' as md;

import 'citations.dart';

/// 行内样式（可叠加的位标志）。
class InlineStyle {
  const InlineStyle(this.bits);

  final int bits;

  static const none = InlineStyle(0);
  static const bold = InlineStyle(1 << 0);
  static const italic = InlineStyle(1 << 1);
  static const code = InlineStyle(1 << 2);
  static const strikethrough = InlineStyle(1 << 3);
  static const highlight = InlineStyle(1 << 4);

  InlineStyle operator |(InlineStyle other) => InlineStyle(bits | other.bits);

  bool has(InlineStyle flag) => (bits & flag.bits) != 0;

  @override
  bool operator ==(Object other) => other is InlineStyle && other.bits == bits;

  @override
  int get hashCode => bits;

  @override
  String toString() => 'InlineStyle($bits)';
}

/// 一段行内文本及其样式；[citation] 非空表示这是一处引用芯片。
class InlineRun {
  const InlineRun({
    required this.text,
    this.style = InlineStyle.none,
    this.link,
    this.citation,
  });

  final String text;
  final InlineStyle style;
  final Uri? link;
  final CitationRef? citation;

  InlineRun copyWith({String? text, InlineStyle? style}) => InlineRun(
    text: text ?? this.text,
    style: style ?? this.style,
    link: link,
    citation: citation,
  );

  @override
  bool operator ==(Object other) =>
      other is InlineRun &&
      other.text == text &&
      other.style == style &&
      other.link == link &&
      other.citation == citation;

  @override
  int get hashCode => Object.hash(text, style, link, citation);

  @override
  String toString() => 'InlineRun($text, ${style.bits})';
}

/// 块级结构。渲染层按此逐块画 widget。
sealed class MarkdownBlock {
  const MarkdownBlock();
}

final class MdHeading extends MarkdownBlock {
  const MdHeading(this.level, this.runs);

  final int level;
  final List<InlineRun> runs;
}

/// [anchor] 来自段首的 `<a id="pN"></a>`，供滚动定位。
final class MdParagraph extends MarkdownBlock {
  const MdParagraph(this.runs, {this.anchor});

  final List<InlineRun> runs;
  final int? anchor;
}

final class MdBlockQuote extends MarkdownBlock {
  const MdBlockQuote(this.blocks);

  final List<MarkdownBlock> blocks;
}

final class MdList extends MarkdownBlock {
  const MdList({
    required this.ordered,
    required this.start,
    required this.items,
  });

  final bool ordered;
  final int start;
  final List<List<MarkdownBlock>> items;
}

final class MdCodeBlock extends MarkdownBlock {
  const MdCodeBlock({this.language, required this.code});

  final String? language;
  final String code;
}

final class MdTable extends MarkdownBlock {
  const MdTable({required this.header, required this.rows});

  final List<List<InlineRun>> header;
  final List<List<List<InlineRun>>> rows;
}

final class MdThematicBreak extends MarkdownBlock {
  const MdThematicBreak();
}

/// 无法识别的 HTML 块，按等宽纯文本显示。
final class MdHtml extends MarkdownBlock {
  const MdHtml(this.raw);

  final String raw;
}

/// 取一组 run 的纯文本（用于引文高亮定位、无障碍标签）。
String plainTextOf(Iterable<InlineRun> runs) => runs.map((r) => r.text).join();

/// 转义方括号占位标签：package:markdown 在 `_combineAdjacentText` 里把相邻文本节点合并，
/// `\[1]` 转义后与后文并成一个 `A[1] B` 文本节点，事后扫描无法再区分转义。
/// 因此用一个自定义行内语法把 `\[` 变成独立元素节点，扫描时不参与标记识别。
const _escapedBracketTag = 'yaoe-escaped-bracket';

class _EscapedBracketSyntax extends md.InlineSyntax {
  _EscapedBracketSyntax() : super(r'\\\[', startCharacter: 0x5c);

  @override
  bool onMatch(md.InlineParser parser, Match match) {
    parser.addNode(md.Element.text(_escapedBracketTag, '['));
    return true;
  }
}

/// 解析 Markdown。[citationLimit] 非空时把 `[n]` / `[n¶pid]` 裸标记识别为引用
/// （1…limit 之外不识别）。
List<MarkdownBlock> parseMarkdown(String markdown, {int? citationLimit}) {
  final document = md.Document(
    // 自定义语法先于默认语法求值，`\[` 因此在 EscapeSyntax 之前被摘成独立节点。
    inlineSyntaxes: [_EscapedBracketSyntax()],
    extensionSet: md.ExtensionSet.gitHubFlavored,
    encodeHtml: false,
  );
  return _blocks(document.parse(markdown), citationLimit);
}

const _blockTags = <String>{
  'p',
  'ul',
  'ol',
  'blockquote',
  'pre',
  'table',
  'hr',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
};

List<MarkdownBlock> _blocks(List<md.Node> nodes, int? limit) {
  final out = <MarkdownBlock>[];
  for (final node in nodes) {
    final block = _block(node, limit);
    if (block != null) out.add(block);
  }
  return out;
}

MarkdownBlock? _block(md.Node node, int? limit) {
  // 块级 HTML 由 HtmlBlockSyntax 直接产出裸 Text 节点。
  if (node is md.Text) {
    final raw = node.text.trim();
    return raw.isEmpty ? null : MdHtml(raw);
  }
  if (node is! md.Element) return null;
  final children = node.children ?? const <md.Node>[];

  switch (node.tag) {
    case 'h1' || 'h2' || 'h3' || 'h4' || 'h5' || 'h6':
      final level = int.tryParse(node.tag.substring(1)) ?? 1;
      return MdHeading(level, _inline(children, limit).runs);
    case 'p':
      final built = _inline(children, limit);
      return MdParagraph(built.runs, anchor: built.anchor);
    case 'blockquote':
      return MdBlockQuote(_blocks(children, limit));
    case 'ul':
      return MdList(ordered: false, start: 1, items: _items(children, limit));
    case 'ol':
      return MdList(
        ordered: true,
        start: int.tryParse(node.attributes['start'] ?? '') ?? 1,
        items: _items(children, limit),
      );
    case 'pre':
      final code = children.whereType<md.Element>().firstWhere(
        (e) => e.tag == 'code',
        orElse: () => md.Element.text('code', node.textContent),
      );
      final classes = code.attributes['class'] ?? '';
      final language = classes.startsWith('language-')
          ? classes.substring('language-'.length)
          : null;
      return MdCodeBlock(
        language: (language ?? '').isEmpty ? null : language,
        code: code.textContent.replaceFirst(RegExp(r'\n$'), ''),
      );
    case 'table':
      final header = <List<InlineRun>>[];
      final rows = <List<List<InlineRun>>>[];
      for (final section in children.whereType<md.Element>()) {
        for (final row in (section.children ?? const <md.Node>[])
            .whereType<md.Element>()
            .where((e) => e.tag == 'tr')) {
          final cells = (row.children ?? const <md.Node>[])
              .whereType<md.Element>()
              .map(
                (cell) =>
                    _inline(cell.children ?? const <md.Node>[], limit).runs,
              )
              .toList(growable: false);
          if (section.tag == 'thead') {
            header.addAll(cells);
          } else {
            rows.add(cells);
          }
        }
      }
      return MdTable(header: header, rows: rows);
    case 'hr':
      return const MdThematicBreak();
    default:
      // 未知块容器：能拆出子块就按引用块渲染，否则当段落。
      final inner = _blocks(children, limit);
      if (inner.isNotEmpty) return MdBlockQuote(inner);
      final built = _inline(children, limit);
      return built.runs.isEmpty
          ? null
          : MdParagraph(built.runs, anchor: built.anchor);
  }
}

List<List<MarkdownBlock>> _items(List<md.Node> children, int? limit) => children
    .whereType<md.Element>()
    .where((e) => e.tag == 'li')
    .map((li) => _itemBlocks(li.children ?? const <md.Node>[], limit))
    .toList(growable: false);

/// 列表项：紧凑列表的子节点是行内节点，需要自己聚成段落。
List<MarkdownBlock> _itemBlocks(List<md.Node> children, int? limit) {
  final blocks = <MarkdownBlock>[];
  final pending = <md.Node>[];

  void flush() {
    if (pending.isEmpty) return;
    final built = _inline(pending, limit);
    if (built.runs.isNotEmpty) {
      blocks.add(MdParagraph(built.runs, anchor: built.anchor));
    }
    pending.clear();
  }

  for (final child in children) {
    if (child is md.Element && _blockTags.contains(child.tag)) {
      flush();
      final block = _block(child, limit);
      if (block != null) blocks.add(block);
    } else {
      pending.add(child);
    }
  }
  flush();
  return blocks;
}

class _InlineBuilder {
  final List<InlineRun> runs = [];
  int? anchor;
  bool highlight = false;

  /// 是否已出现可见文本（决定 `<a id="pN">` 是否算段落锚点）。
  bool sawText = false;

  void add(String text, InlineStyle style, Uri? link, CitationRef? citation) {
    if (text.isEmpty) return;
    runs.add(
      InlineRun(
        text: text,
        style: highlight ? style | InlineStyle.highlight : style,
        link: link,
        citation: citation,
      ),
    );
    sawText = true;
  }
}

_InlineBuilder _inline(List<md.Node> children, int? limit) {
  final builder = _InlineBuilder();
  for (final child in children) {
    _appendInline(
      child,
      builder,
      style: InlineStyle.none,
      link: null,
      allowMarkers: true,
      limit: limit,
    );
  }
  return builder;
}

void _appendInline(
  md.Node node,
  _InlineBuilder builder, {
  required InlineStyle style,
  required Uri? link,
  required bool allowMarkers,
  required int? limit,
}) {
  if (node is md.Text) {
    _appendText(
      node.text,
      builder,
      style: style,
      link: link,
      allowMarkers: allowMarkers,
      limit: limit,
    );
    return;
  }
  if (node is! md.Element) return;
  final children = node.children ?? const <md.Node>[];

  void recurse(InlineStyle nextStyle, Uri? nextLink, bool markers) {
    for (final child in children) {
      _appendInline(
        child,
        builder,
        style: nextStyle,
        link: nextLink,
        allowMarkers: markers,
        limit: limit,
      );
    }
  }

  switch (node.tag) {
    case 'em':
      recurse(style | InlineStyle.italic, link, allowMarkers);
    case 'strong':
      recurse(style | InlineStyle.bold, link, allowMarkers);
    case 'del':
      recurse(style | InlineStyle.strikethrough, link, allowMarkers);
    case 'mark':
      builder.highlight = true;
      recurse(style, link, allowMarkers);
      builder.highlight = false;
    case 'code':
      // 代码里的方括号不是引用标记。
      builder.add(node.textContent, style | InlineStyle.code, link, null);
    case 'a':
      final destination = node.attributes['href'] ?? '';
      final ref = parseAnswerLink(destination);
      if (ref != null) {
        builder.add(node.textContent, style, null, ref);
      } else {
        // 普通链接的标签不再识别裸标记：`[[1]](https://example.com)` 必须留在原链接上，
        // 否则点击会跳到第 1 篇文献（与 Web 端跳过整个 link 子树一致）。
        recurse(style, Uri.tryParse(destination) ?? link, false);
      }
    case 'img':
      final alt = node.attributes['alt'] ?? node.attributes['title'] ?? '';
      builder.add(alt, style, link, null);
    case 'br':
      builder.add('\n', style, link, null);
    case _escapedBracketTag:
      // 转义出来的方括号是普通字符，不参与引用标记识别。
      builder.add(node.textContent, style, link, null);
    default:
      recurse(style, link, allowMarkers);
  }
}

final _anchorRegex = RegExp(r'<a\s+id="p(\d+)"\s*>', caseSensitive: false);
final _inlineTagRegex = RegExp(
  r'<a\s+id="p\d+"\s*>|</a\s*>|<mark\s*>|</mark\s*>|<br\s*/?>',
  caseSensitive: false,
);

/// 文本节点：切出行内 HTML（段落锚点、`<mark>` 开关、`<br>`）与引用标记。
///
/// package:markdown 的 `InlineHtmlSyntax` 把原始标签留在普通文本里（不单独成节点），
/// 因此这里必须自己扫描；转义 `\[1]` 由 `EscapeSyntax` 拆成独立文本节点，天然不成标记。
void _appendText(
  String text,
  _InlineBuilder builder, {
  required InlineStyle style,
  required Uri? link,
  required bool allowMarkers,
  required int? limit,
}) {
  var cursor = 0;
  for (final match in _inlineTagRegex.allMatches(text)) {
    if (match.start > cursor) {
      _appendPlain(
        text.substring(cursor, match.start),
        builder,
        style: style,
        link: link,
        allowMarkers: allowMarkers,
        limit: limit,
      );
    }
    final tag = match[0]!.toLowerCase();
    if (tag.startsWith('<a')) {
      if (!builder.sawText) {
        builder.anchor = int.tryParse(
          _anchorRegex.firstMatch(match[0]!)?.group(1) ?? '',
        );
      }
    } else if (tag.startsWith('<mark')) {
      builder.highlight = true;
    } else if (tag.startsWith('</mark')) {
      builder.highlight = false;
    } else if (tag.startsWith('<br')) {
      builder.add('\n', style, link, null);
    }
    cursor = match.end;
  }
  if (cursor < text.length) {
    _appendPlain(
      text.substring(cursor),
      builder,
      style: style,
      link: link,
      allowMarkers: allowMarkers,
      limit: limit,
    );
  }
}

void _appendPlain(
  String text,
  _InlineBuilder builder, {
  required InlineStyle style,
  required Uri? link,
  required bool allowMarkers,
  required int? limit,
}) {
  // 软换行在行内等价于空格（硬换行由 `br` 元素单独产出 '\n'）。
  final flowed = text.replaceAll('\n', ' ');
  if (!allowMarkers || limit == null || limit < 1) {
    builder.add(flowed, style, link, null);
    return;
  }
  var cursor = 0;
  for (final marker in markersIn(flowed)) {
    if (marker.ref.n < 1 || marker.ref.n > limit) continue;
    if (marker.start > cursor) {
      builder.add(flowed.substring(cursor, marker.start), style, link, null);
    }
    builder.add(
      flowed.substring(marker.start, marker.end),
      style,
      link,
      marker.ref,
    );
    cursor = marker.end;
  }
  if (cursor < flowed.length) {
    builder.add(flowed.substring(cursor), style, link, null);
  }
}

File: apple/PicoSeekKit/Sources/PicoSeekKit/Logic/MarkdownDocument.swift
import Foundation
import Markdown

/// 行内样式（可叠加）。
public struct InlineStyle: OptionSet, Sendable, Hashable {
    public let rawValue: Int

    public init(rawValue: Int) {
        self.rawValue = rawValue
    }

    public static let bold = InlineStyle(rawValue: 1 << 0)
    public static let italic = InlineStyle(rawValue: 1 << 1)
    public static let code = InlineStyle(rawValue: 1 << 2)
    public static let strikethrough = InlineStyle(rawValue: 1 << 3)
    public static let highlight = InlineStyle(rawValue: 1 << 4)
}

/// 一段行内文本及其样式；`citation` 非空表示这是一处引用芯片。
public struct InlineRun: Hashable, Sendable {
    public var text: String
    public var style: InlineStyle
    public var link: URL?
    public var citation: CitationRef?

    public init(text: String, style: InlineStyle = [], link: URL? = nil, citation: CitationRef? = nil) {
        self.text = text
        self.style = style
        self.link = link
        self.citation = citation
    }
}

/// 块级结构。渲染层按此逐块画 SwiftUI 视图。
public indirect enum MarkdownBlock: Hashable, Sendable, Identifiable {
    case heading(level: Int, runs: [InlineRun])
    /// `anchor` 来自段首的 `<a id="pN"></a>`，供滚动定位。
    case paragraph(runs: [InlineRun], anchor: Int?)
    case blockQuote([MarkdownBlock])
    case list(ordered: Bool, start: Int, items: [[MarkdownBlock]])
    case codeBlock(language: String?, code: String)
    case table(header: [[InlineRun]], rows: [[[InlineRun]]])
    case thematicBreak
    /// 无法识别的 HTML 块，按等宽纯文本显示。
    case html(String)

    public var id: Int { hashValue }
}

public enum MarkdownDocument {
    /// 解析 Markdown。`citationLimit` 非空时把 `[n]` / `[n¶pid]` 裸标记识别为引用（1...limit 之外不识别）。
    public static func parse(_ markdown: String, citationLimit: Int?) -> [MarkdownBlock] {
        let context = Context(source: markdown, citationLimit: citationLimit)
        return blocks(from: Document(parsing: markdown).children, context)
    }

    /// 取一组 run 的纯文本（用于引文高亮定位、无障碍标签）。
    public static func plainText(of runs: [InlineRun]) -> String {
        runs.map(\.text).joined()
    }

    // MARK: - 解析上下文

    /// 缓存源码字节与行首偏移：cmark 的列号是 1-based 的 UTF-8 字节偏移。
    private struct Context {
        let bytes: [UInt8]
        let lineStarts: [Int]
        let citationLimit: Int?

        init(source: String, citationLimit: Int?) {
            let bytes = Array(source.utf8)
            var starts = [0]
            for (offset, byte) in bytes.enumerated() where byte == UInt8(ascii: "\n") {
                starts.append(offset + 1)
            }
            self.bytes = bytes
            lineStarts = starts
            self.citationLimit = citationLimit
        }

        /// 节点对应的原始源码片段（保留转义反斜杠）。
        func rawText(_ range: SourceRange?) -> String? {
            guard let range,
                  range.lowerBound.line >= 1, range.lowerBound.line <= lineStarts.count,
                  range.upperBound.line >= 1, range.upperBound.line <= lineStarts.count
            else { return nil }
            let start = lineStarts[range.lowerBound.line - 1] + range.lowerBound.column - 1
            let end = lineStarts[range.upperBound.line - 1] + range.upperBound.column - 1
            guard start >= 0, end <= bytes.count, start <= end else { return nil }
            return String(decoding: bytes[start ..< end], as: UTF8.self)
        }
    }

    // MARK: - 块级

    private static func blocks(from children: some Sequence<any Markup>, _ context: Context) -> [MarkdownBlock] {
        children.compactMap { block(from: $0, context) }
    }

    private static func block(from markup: any Markup, _ context: Context) -> MarkdownBlock? {
        switch markup {
        case let heading as Heading:
            return .heading(level: heading.level, runs: inlineRuns(heading.children, context).runs)
        case let paragraph as Markdown.Paragraph:
            let built = inlineRuns(paragraph.children, context)
            return .paragraph(runs: built.runs, anchor: built.anchor)
        case let quote as BlockQuote:
            return .blockQuote(blocks(from: quote.children, context))
        case let list as UnorderedList:
            return .list(ordered: false, start: 1, items: list.listItems.map { blocks(from: $0.children, context) })
        case let list as OrderedList:
            return .list(ordered: true, start: Int(list.startIndex), items: list.listItems.map { blocks(from: $0.children, context) })
        case let code as CodeBlock:
            return .codeBlock(language: code.language, code: code.code)
        case let table as Markdown.Table:
            let header = Array(table.head.cells.map { inlineRuns($0.children, context).runs })
            let rows = Array(table.body.rows.map { row in Array(row.cells.map { inlineRuns($0.children, context).runs }) })
            return .table(header: header, rows: rows)
        case is ThematicBreak:
            return .thematicBreak
        case let html as HTMLBlock:
            let text = html.rawHTML.trimmingCharacters(in: .whitespacesAndNewlines)
            return text.isEmpty ? nil : .html(text)
        case let container as any BlockContainer:
            let inner = blocks(from: container.children, context)
            return inner.isEmpty ? nil : .blockQuote(inner)
        default:
            return nil
        }
    }

    // MARK: - 行内

    private struct InlineBuilder {
        var runs: [InlineRun] = []
        var anchor: Int?
        var highlight = false
        /// 段首之前是否已经出现过可见文本（决定 `<a id="pN">` 是否算段落锚点）。
        var sawText = false

        mutating func append(_ text: String, style: InlineStyle, link: URL?, citation: CitationRef?) {
            guard !text.isEmpty else { return }
            var style = style
            if highlight { style.insert(.highlight) }
            runs.append(InlineRun(text: text, style: style, link: link, citation: citation))
            sawText = true
        }
    }

    private static func inlineRuns(_ children: some Sequence<any Markup>, _ context: Context) -> InlineBuilder {
        var builder = InlineBuilder()
        for child in children {
            appendInline(child, style: [], link: nil, into: &builder, context)
        }
        return builder
    }

    private static func appendInline(
        _ markup: any Markup,
        style: InlineStyle,
        link: URL?,
        allowMarkers: Bool = true,
        into builder: inout InlineBuilder,
        _ context: Context
    ) {
        switch markup {
        case let text as Markdown.Text:
            appendText(text, style: style, link: link, allowMarkers: allowMarkers, into: &builder, context)

        case let emphasis as Emphasis:
            for child in emphasis.children {
                appendInline(child, style: style.union(.italic), link: link, allowMarkers: allowMarkers, into: &builder, context)
            }

        case let strong as Strong:
            for child in strong.children {
                appendInline(child, style: style.union(.bold), link: link, allowMarkers: allowMarkers, into: &builder, context)
            }

        case let struck as Strikethrough:
            for child in struck.children {
                appendInline(child, style: style.union(.strikethrough), link: link, allowMarkers: allowMarkers, into: &builder, context)
            }

        case let code as InlineCode:
            // 代码里的方括号不是引用标记。
            builder.append(code.code, style: style.union(.code), link: link, citation: nil)

        case let inlineLink as Markdown.Link:
            let destination = inlineLink.destination ?? ""
            let text = inlineLink.plainText
            if let ref = Citations.parseAnswerLink(destination) {
                builder.append(text, style: style, link: nil, citation: ref)
            } else {
                // 普通链接的标签不再识别裸标记：`[[1]](https://example.com)` 必须留在原链接上，
                // 否则点击会跳到第 1 篇文献（与 Web 端跳过整个 link 子树一致）。
                let url = URL(string: destination)
                for child in inlineLink.children {
                    appendInline(child, style: style, link: url ?? link, allowMarkers: false, into: &builder, context)
                }
            }

        case let image as Markdown.Image:
            let alt = image.plainText.isEmpty ? (image.title ?? "") : image.plainText
            builder.append(alt, style: style, link: link, citation: nil)

        case is SoftBreak:
            builder.append(" ", style: style, link: link, citation: nil)

        case is LineBreak:
            builder.append("\n", style: style, link: link, citation: nil)

        case let html as InlineHTML:
            appendInlineHTML(html.rawHTML, into: &builder)

        case let container as any InlineContainer:
            for child in container.children {
                appendInline(child, style: style, link: link, allowMarkers: allowMarkers, into: &builder, context)
            }

        default:
            break
        }
    }

    /// 段落锚点 `<a id="pN">`、`</a>`、`<mark>` 开关；其余行内 HTML 丢弃。
    private static func appendInlineHTML(_ rawHTML: String, into builder: inout InlineBuilder) {
        let html = rawHTML.trimmingCharacters(in: .whitespaces)
        if !builder.sawText, let anchor = anchorID(in: html) {
            builder.anchor = anchor
            return
        }
        switch html.lowercased() {
        case "<mark>": builder.highlight = true
        case "</mark>": builder.highlight = false
        default: break
        }
    }

    private static let anchorRegex = try! NSRegularExpression(pattern: #"^<a\s+id="p(\d+)"\s*>$"#, options: [.caseInsensitive])

    private static func anchorID(in html: String) -> Int? {
        let range = NSRange(html.startIndex ..< html.endIndex, in: html)
        guard let match = anchorRegex.firstMatch(in: html, range: range),
              let group = Range(match.range(at: 1), in: html) else { return nil }
        return Int(html[group])
    }

    /// 文本节点：按需要切出引用标记。
    private static func appendText(
        _ node: Markdown.Text,
        style: InlineStyle,
        link: URL?,
        allowMarkers: Bool,
        into builder: inout InlineBuilder,
        _ context: Context
    ) {
        guard allowMarkers, let limit = context.citationLimit, limit >= 1 else {
            builder.append(node.string, style: style, link: link, citation: nil)
            return
        }

        // cmark 的行内节点在缩进段落、续行等情形下列号会偏移，实体引用也会让源码与解析结果不等长。
        // 因此只有「反转义后与解析结果完全一致」的源码片段才可信；否则退回解析后的文本，
        // 宁可把 `&#91;1]` 这种实体拼出来的方括号当成引用，也绝不拿错位切片重建正文（会丢字）。
        let raw = context.rawText(node.range)
        let trustedRaw = raw.flatMap { unescape($0) == node.string ? $0 : nil }
        let text = trustedRaw ?? node.string

        var cursor = text.startIndex
        for marker in Citations.markers(in: text) {
            guard marker.ref.n >= 1, marker.ref.n <= limit else { continue }
            // 只有可信源码才能判断转义：`\[1]` 不是引用。
            if trustedRaw != nil, marker.range.lowerBound > text.startIndex,
               text[text.index(before: marker.range.lowerBound)] == "\\" { continue }
            if cursor < marker.range.lowerBound {
                let segment = String(text[cursor ..< marker.range.lowerBound])
                builder.append(trustedRaw == nil ? segment : unescape(segment), style: style, link: link, citation: nil)
            }
            builder.append(String(text[marker.range]), style: style, link: link, citation: marker.ref)
            cursor = marker.range.upperBound
        }
        if cursor < text.endIndex {
            let segment = String(text[cursor...])
            builder.append(trustedRaw == nil ? segment : unescape(segment), style: style, link: link, citation: nil)
        }
    }

    /// 还原 CommonMark 的反斜杠转义（仅 ASCII 标点可被转义）。
    private static func unescape(_ text: String) -> String {
        var out = ""
        out.reserveCapacity(text.count)
        var escaping = false
        for character in text {
            if escaping {
                if !character.isASCII || !character.isPunctuation && !character.isSymbol { out.append("\\") }
                out.append(character)
                escaping = false
            } else if character == "\\" {
                escaping = true
            } else {
                out.append(character)
            }
        }
        if escaping { out.append("\\") }
        return out
    }
}
